llvm/lib/CAS/ObjectStore.cpp

```diff
@@ -198,22 +198,60 @@ ObjectProxy::getMemoryBuffer(StringRef Name,
   return CAS->getMemoryBuffer(H, Name, RequiresNullTerminator);
 }
 
-static Expected<std::unique_ptr<ObjectStore>>
+static Expected<std::shared_ptr<ObjectStore>>
+createOnDiskCASImpl(const Twine &Path) {
+  return createOnDiskCAS(Path);
+}
+
+static Expected<std::shared_ptr<ObjectStore>>
 createInMemoryCASImpl(const Twine &) {
   return createInMemoryCAS();
 }
 
+static Expected<std::shared_ptr<ObjectStore>>
+createPluginCASImpl(const Twine &URL) {
+  // Format used is
+  //   plugin://${PATH_TO_PLUGIN}?${OPT1}=${VAL1}&${OPT2}=${VAL2}..
+  // "ondisk-path" as option is treated specially, the rest of options are
+  // passed to the plugin verbatim.
+  SmallString<256> PathBuf;
+  auto [PluginPath, Options] = URL.toStringRef(PathBuf).split('?');
+  std::string OnDiskPath;
+  SmallVector<std::pair<std::string, std::string>> PluginArgs;
+  while (!Options.empty()) {
+    StringRef Opt;
+    std::tie(Opt, Options) = Options.split('&');
+    auto [Name, Value] = Opt.split('=');
+    if (Name == "ondisk-path") {
```

Comment:
I prefer the option processing to be moved into the plugin. In this function, it should validate the general structure of the URL, create a `std::vector<std::string>` and pass that to the plugin. The plugin can decide if that is a flag or an option or if use `=` as separator.

I guess the only problem for using a more general pattern described above is there seems to be a requirement for sharing an OnDiskPath. If that is a requirement, maybe we can add a callback for the plugin to report that back.

Reply:
> I prefer the option processing to be moved into the plugin.

What is the benefit? Currently the plugin has no need to parse options, it receives them as a name/value pair.
If you are concerned about code sharing this can move to some common function, but I don't see a particular need to add a C API for option parsing.

Reply:
Just a better design for plugin. No specific benefit for existing implementation.

Reply:
I don't like requiring each plugin implementation to implement their own option parsing, seems unnecessary requirement.

Reply:
I don't think asking plugin to implement a `std::vector<std::string>> Args` interface is harder than `std::vector<std::pair<std::string, std::string>>>`. The later locked the format into a key/value pair, not a flag.

My main complaint about the current interface is that you have a CASPath parameter which has special meaning. Maybe the better way is to have a C API to ask plugin to pass that shared path back based on the parameter on the command line. Just a complaint, not an objection to current implementation.

Reply:
For the plugin API there is explicit API to pass a local path:
```
/**
 * Receives a local file-system path that the plugin should use for any on-disk
 * resources/caches.
 */
LLCASAPI_PUBLIC void
llcasapi_cas_options_set_ondisk_path(llcasapi_cas_options_t, const char *path);
```

The rationale to treat a local path for on-disk caches specially is that this is an important property that almost every plugin would use (if a plugin doesn't have a use for it it can just ignore it).

Having a structured way to pass the local path simplifies clients, like a build system, so they can have a consistent way to pass a path for local caches for any plugin, without having to figure out what special name for the option the plugin accepts.

Reply:
To clarify a bit, I don't consider the "plugin://" format/interface important here, note that it is only used for the testing tools.
I don't want to expose it as an interface between clang and the build system or any other tool that would be used by the build system.